src/vis/vis.js

```diff
@@ -509,7 +516,47 @@ var Vis = View.extend({
       }
     });
 
-    // TODO: We can probably move this logic somewhere else
+    // Initialize widgets
+    var hasLayersWithWidgets = this._initializeWidgets(data, interactiveLayers, cartoDBLayerGroup);
+
+    this.map.layers.reset(layers);
+    this.overlayModels.reset(data.overlays);
+
+    // if there are no sublayer_options fill it
+    if (!options.sublayer_options) {
+      this._setupSublayers(data.layers, options);
+    }
+
+    this._setLayerOptions(options);
+
+    _.defer(function() {
+      self.trigger('done', self, map.layers);
+    });
+
+    // TODO: rethink this, it is only a hack
+    if (hasLayersWithWidgets) {
+      setTimeout(function() {
+        self.mapView.invalidateSize();
+      }, 0);
+    }
+
+    return this;
+  },
+
+  /**
+   * Initializes the widgets related classes.
+   *
+   * @private
+   * @param  {Object} vizconfig         Object with viz data configuration.
+   * @param  {Array.<Object>} interactiveLayers Array with references to
+   * the layers withint the viz config.
+   * @param  {(CartoDBLayerGroupNamed|CartoDBLayerGroupAnonymous)} cartoDBLayerGroup Layer group reference.
+   * @return {boolean}  True if any layer contains widgets and has been initialized,
+   * false otherwise.
+   */
+  _initializeWidgets: function(vizconfig, interactiveLayers, cartoDBLayerGroup) {
```

Comment:
This method is doing much more than initialising the widgets (no single responsibility here)... don't you think?  For example, it's creating the `DashboardInfoView` which has nothing to do with widgets. So it's hard for me to know what it does just by looking at it's name, and also to guess it returns a boolean (methods that return a boolean usually looking more like `_isSomething`, `_hasSomething`, `_isASomething`, etc.)